File: Array/Anubha/boatsToSavePeople.java
import java.util.Arrays;

public class boatsToSavePeople {
    public static void main(String[] args) {
        int[] people = {1,1,1,2};
        int limit = 3;
        System.out.println(numRescueBoats(people, limit));
    }

    public static int numRescueBoats(int[] people, int limit) {
        int boats = 0;
        Arrays.sort(people);
        int i = 0;
        int j = people.length - 1;
        while (i <= j) {
            if (people[i] + people[j] <= limit) {
                i++;
                j--;
                boats++;
            } else {
                j--;
                boats++;
            }
        }
        return boats;
    }
}

File: Array/Shefali/Boats_to_Save_People.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        sort(people.begin(),people.end());
        int boats=0;
        int left=0;
        int right=people.size()-1;
        while(left<=right){
            int sum=people[left]+people[right];
            if(sum<=limit){
                left++;
                right--;
            }
            else{        // sum exceeds the limit, then heaviest person will go alone, hence updating only the right pointer
                right--;
            }
            boats++;
        }
        return boats;
    }
};
